File: 17. Letter Combinations of a Phone Number.py
class Solution:
    digs_to_letts = {"1": "*",
                     "2": "abc",
                     "3": "def",
                     "4": "ghi",
                     "5": "jkl",
                     "6": "mno",
                     "7": "pqrs",
                     "8": "tuv",
                     "9": "wxyz"}
    
    def letterCombinations(self, digits: 'str') -> 'List[str]':
        
        num_len = len(digits)
        if num_len == 0 or not digits:
            return []
        
        if num_len == 1:
            return list(self.digs_to_letts[digits])
        else:
            result = []
            for l in self.digs_to_letts[digits[0]]:
                result += [l+letters for letters in self.letterCombinations(digits[1:])]
            return result
        

File: 17. Letter Combinations of a Phone Number/17. Letter Combinations of a Phone Number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
    const digit_map = {"1": "*",
                       "2": "abc",
                       "3": "def",
                       "4": "ghi",
                       "5": "jkl",
                       "6": "mno",
                       "7": "pqrs",
                       "8": "tuv",
                       "9": "wxyz"}
    
    let queue = digits.length === 0 ? []:[""];
    let letters, queueLength, itemFromQueue;
    
    for (var i = 0; i < digits.length; i++) {
        letters = digit_map[digits[i]];
        queueLength = queue.length;
        
        for (var j = 0; j < queueLength; j++) {
            itemFromQueue = queue.shift();
            for (var k = 0; k < letters.length; k++) {
                queue.push(itemFromQueue + letters[k])
            }
        }
    }
    return queue;
};

